packages/lib/server/repository/team.ts

```diff
@@ -0,0 +1,149 @@
+import type { Prisma } from "@prisma/client";
+import type { z } from "zod";
+
+import { whereClauseForOrgWithSlugOrRequestedSlug } from "@calcom/ee/organizations/lib/orgDomains";
+import logger from "@calcom/lib/logger";
+import prisma from "@calcom/prisma";
+import { teamMetadataSchema } from "@calcom/prisma/zod-utils";
+
+type TeamGetPayloadWithParsedMetadata<TeamSelect extends Prisma.TeamSelect> =
+  | (Omit<Prisma.TeamGetPayload<{ select: TeamSelect }>, "metadata"> & {
+      metadata: z.infer<typeof teamMetadataSchema>;
+    })
+  | null;
+
+type GetTeamOrOrgArg<TeamSelect extends Prisma.TeamSelect> = {
+  lookupBy: (
+    | {
+        id: number;
+      }
+    | {
+        slug: string;
+      }
+  ) & {
+    havingMemberWithId?: number;
+  };
+  /**
+   * If we are fetching a team, this is the slug of the organization that the team belongs to.
+   */
+  forOrgWithSlug: string | null;
+  /**
+   * If true, means that we need to fetch an organization with the given slug. Otherwise, we need to fetch a team with the given slug.
+   */
+  isOrg: boolean;
+  teamSelect: TeamSelect;
+};
+
+const log = logger.getSubLogger({ prefix: ["repository", "team"] });
+
+/**
+ * Get's the team or organization with the given slug or id reliably along with parsed metadata.
+ */
+async function getTeamOrOrg<TeamSelect extends Prisma.TeamSelect>({
```

Comment:
Prisma doesn't make it easy to wrap generics over it.

Reply:
Do we actually need a generic here? Or can we we just use Prisma.TeamSelect as the type of team select

I made an example that seems to work - see your DMs
https://threads.com/messages/ef1928b9-ab1a-4903-bad2-670138a519e6?messageID=d40dff2c-de71-47ae-ab38-7f3770452777

Reply:
Discussed with Sean. It won't work